[lld] Warn for missing needed sections by --gdb-index

aury6623 opened the pull request:
Issue a warning when lld is passed `--gdb-index` and some object files have general debug info but not the specific debug  sections generated by `-ggnu-pubnames` or `-gsplit-dwarf`. Missing those sections on all inputs leads to an empty index, which gdb ignores. Missing those sections on only some sources leads to a partially complete index, which confuses gdb.

Addresses https://github.com/llvm/llvm-project/issues/156732 and https://github.com/llvm/llvm-project/issues/34168

aury6623 committed:
[lld] Warn when missing needed sections for --gdb-index

lenary reviewed (commented):
Please can you add a test?

aury6623 committed:
add new test for warning, update existing test to not fail with the warning

aury6623 commented:
@lenary Thanks for the help! I added a test. I also updated an existing test which I realized was failing because it was expecting no output from ld.lld but was now getting the warning. I wasn't sure if adding those the missing sections to that test could impact the test's intent, so I figured the best thing to do is to just make the test ignore that particular warning by stripping it out with sed.

lenary commented:
Thanks for adding the test.

I think using `sed` isn't great in the pipeline, as I don't know if it's available on Windows.

I think it would be fine to update that test to also check for the warning, but I'm not entirely sure.

MaskRay reviewed (changes requested):
Thanks for the motivation to improve things but I believe this will certainly break some corporate users. Left a comment on #34168  . 

> While warnings are generally acceptable in Linux distros and OSS environments, corporate users and large projects using custom toolchains often upgrade warnings to errors by specifying --fatal-warnings. Reporting a warning should probably be deferred to a lint tool. It's also quite difficult to ensure that all object files have .debug_gnu_pubnames - when you have prebuilt object files.

aury6623 commented:
@lenary 

> I think using sed isn't great in the pipeline, as I don't know if it's available on Windows.

I was wondering that too, but I checked to see if other tests use sed, and found lots of them. Here's one in the same directory which only requires x86 and uses sed:

https://github.com/llvm/llvm-project/blob/1c196452dd4afd90f0d239a6d9aeb31b136e3df4/lld/test/ELF/debug-names-different-aug-string.s#L1-L3

and here's one from llvm-objdump which specifically requires windows but still uses sed:

https://github.com/llvm/llvm-project/blob/39fbec06222ecb37b489d69b45d801efa264bf26/llvm/test/tools/llvm-objdump/X86/source-interleave-prefix-windows.test#L2-L7

So it _seems_ fine to me.


@MaskRay 
> Thanks for the motivation to improve things but I believe this will certainly break some corporate users.

For what it's worth, I am a corporate user, working towards enabling fatal warnings on our compilations :)
In this case, the warning is really indicating that `--gdb-index` won't work as intended. Having this warning would have saved me a few days of debugging the tools.

That said, I understand a warning could cause trouble here, so what about making it a remark/info message instead of a warning? I just want the tool to provide some feedback that it's not doing what the user probably intended.

> Latest gdb has .debug_names support. And I've found this statement "I consider .gdb_index to be near the end of its life."

I was unable to get the latest gdb to work with DWARF-5 debugging info and C macro debugging, but maybe I was doing something wrong. And yeah, .gdb_index is nearing end of life, but it's not dead yet, and plus a lot of people will continue using old tools for a few years to come. I know my corporate dev environment is still running a very old version of gdb by default unless you decide to build it for yourself. In fairness, running old gdb but using a modern ld.lld with this warning is probably an unusual combo, but it's the situation I'm in.

aury6623 commented:
> what do you think about something opt-in like -z gdb-report={none,warning,error} as we do for a lot of architectural features?

I'm not sure anyone would ever know to look for flag like that, if it's opt-in. I certainly wouldn't have known.


---

@MaskRay
> Latest gdb has .debug_names support. And I've found this statement "I consider .gdb_index to be near the end of its life."

@lenary
> I don't know the next steps forward for the issue you're facing with trying to get gdb working with dwarf 5 and C macros.

Right, I had forgotten the exact problem that I encountered when trying dwarf-5 with gdb. **The problem is that gdb only supports `.debug_names` when generated by gdb itself** (e.g. the `gdb-add-index` tool. Also possibly by gcc/GNU ld, but I'm not sure, it looks like they're still [working on that](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=99178))

You can see in the lastest gdb sources right now, it ignores `.debug_names` unless it was generated by gdb.

https://gitlab.com/gnutools/binutils-gdb/-/blob/fd3b1c4f4c111444dfe6ff9c3f00a5469e7cd8a6/gdb/dwarf2/read-debug-names.c#L626-630

That's the message I see when I try it myself.

Here's the gdb bug ticket for supporting `.debug_names` from clang:
https://sourceware.org/bugzilla/show_bug.cgi?id=32616

I'd love to use standardized dwarf-5 features, but clearly they're not quite ready yet and I don't think we should dismiss .gdb_index as near end of life.

MaskRay commented:
If we have to add a diagnostic, an option like `-z gdb-index-report={none,warning,error}` is probably the way to go.
I'm concerned that this would not be general enough and users would complain that they have -g object files not built with the toolchain, leading to warnings. And then they would like more options like `--gdb-index-report-exclude=a.o`.
If you are a corporate user and you care much about debug info quality, you likely have following steps that process debug info (debuginfod/SymbolStore/...), and you can a debug info lint post linking.

aury6623 commented:
Sorry for the delayed response.
I don't think that having this warning hidden behind a flag that no one knows about would be useful. Probably no one will discover the flag without already knowing what the problem is!
So I'll just close this PR then.
